File: R/Scrips/02.01.Flights.R
#Data Visualization dataset NYCFLIGHTS13

#Clean Environment
rm(list = ls())

# Lod libraries
library(nycflights13)
library(tidyverse)

#Filter example: we can select all flights on January 1st with
filter(flights, month == 1, day == 1)
filter(flights, month == 11 | month == 12)

nov_dec <- filter(flights, month %in% c(11, 12))

#if you wanted to find flights that weren't delayed (on arrival or departure) by 
#more than two hours, you could use:
filter(flights, arr_delay <= 120, dep_delay <= 120)

arrange(flights, year, month, day)
arrange(flights, desc(dep_delay))

flights %>%
  filter(month == 1, day == 1) %>%
  arrange(desc(dep_delay))

# Select columns by name
select(flights, year, month, day)

summarise(flights, delay = mean(dep_delay, na.rm = TRUE))

by_day <- group_by(flights, year, month, day)
summarise(by_day, delay = mean(dep_delay, na.rm = TRUE))


delays <- flights %>% 
  group_by(dest) %>% 
  summarise(count = n(),dist = mean(distance, na.rm = TRUE),
            delay = mean(arr_delay, na.rm = TRUE)) %>% 
  filter(count > 20, dest != "HNL")

ggplot(data = delays, mapping = aes(x = dist, y = delay)) +
  geom_point(aes(size = count), alpha = 1/4) +
  geom_smooth(se = FALSE)

File: Basic Programming/Python/Scripts/02.01.Flights.py
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# 1. Use the 'r' object provided by reticulate
try:
    from src.reticulate import r
except ImportError:
    # If the import fails, we use the built-in 'r' object already in the session
    pass 

# Grab the data from R
flights = r.flights

# ---------------------------------------------------------
# Part I: Filtering & Arranging
# ---------------------------------------------------------

# Select Jan 1st
jan1 = flights.query('month == 1 and day == 1')

# Nov or Dec
nov_dec = flights[flights['month'].isin([11, 12])]

# Not delayed > 2 hours
not_delayed = flights.query('arr_delay <= 120 and dep_delay <= 120')

# Arrange
flights_sorted = flights.sort_values(['year', 'month', 'day'])
flights_delayed_desc = flights.sort_values('dep_delay', ascending=False)

# ---------------------------------------------------------
# Part II: Selection & Summarization
# ---------------------------------------------------------

subset = flights[['year', 'month', 'day']]
avg_delay = flights['dep_delay'].mean()

# The grouping logic
delays = (
    flights.groupby('dest')
    .agg(
        count=('dest', 'size'),
        dist=('distance', 'mean'),
        delay=('arr_delay', 'mean')
    )
    .query('count > 20 and dest != "HNL"')
    .reset_index()
)

# ---------------------------------------------------------
# Part III: Visualization
# ---------------------------------------------------------

sns.set_theme(style="white")
plt.figure(figsize=(10, 6))

sns.regplot(
    data=delays, 
    x='dist', 
    y='delay', 
    lowess=True, 
    scatter_kws={'s': delays['count'] * 0.5, 'alpha': 0.25},
    line_kws={'color': 'blue'}
)

plt.title("Average Arrival Delay vs. Distance")
plt.xlabel("Distance")
plt.ylabel("Arrival Delay")
plt.show()
